### ISSUE TITLE
basic tests

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1402]
Remaining tasks from #1391:
- [ ] distribute the basic tests
- [ ] modify the error function to print the correct address of a basic test that fails
- [ ] possibly modify the locate function in actors4.d to do the above instead
- [ ] reapply https://github.com/Macaulay2/M2/commit/af86e51d499aed8a82966fd9b617419fa201dc2a

**Comment by mahrud** [comment]
@DanGrayson what's the main purpose of the basictests? For certain important things, we have tests in `startup.m2.in` itself, like here for regex:
https://github.com/Macaulay2/M2/blob/9f0ec406451ffa4a1e8141931d1b2a713940131f/M2/Macaulay2/m2/startup.m2.in#L260-L267
Then we have basictests for regex:
https://github.com/Macaulay2/M2/blob/9f0ec406451ffa4a1e8141931d1b2a713940131f/M2/Macaulay2/m2/basictests/regex.m2#L3-L17
But we also have tests for regex here:
https://github.com/Macaulay2/M2/blob/9f0ec406451ffa4a1e8141931d1b2a713940131f/M2/Macaulay2/tests/normal/regex.m2#L1-L17

This is kind of confusing. Would it be good to reduce basictests and move them to `Macaulay2/tests` perhaps? Eventually we could remove the `--check` option as well.

**Comment by DanGrayson** [comment]
The point of running those two batches of tests at startup time is to detect problems in libraries that might prevent startup and thus present mysterious debugging dilemmas.  It would be a good idea to keep those where they are.

**Comment by mahrud** [comment]
Should I move the few test lines in startup.m2.in to basictests as well?

**Comment by DanGrayson** [comment]
No, those are the ones that are important to run before startup, so startup will work.  The ones in basictests are incorporated into the binary so they can also be run at startup time, via `--check 1` on the command line, so they might be helpful, too, but I don't remember having to use that facility.  

Hmm, I don't see that the basictests are run by `make check` -- they should be.  Since they are tests, they could be moved from `m2` to `tests`.

**Comment by mahrud** [comment]
> Hmm, I don't see that the basictests are run by `make check` -- they should be. Since they are tests, they could be moved from `m2` to `tests`.

A minor issue with this is that the `tests` directory is not distributed, and it's kind of unhelpful to run `M2 --check 1` and see an error in a file that doesn't exist. `m2/basictests` isn't currently distributed either, but I think it's small enough that it could be easily distributed under the `share/Macaulay2/Core` directory.

**Comment by mahrud** [comment]
But also, like you said, I've never had to use `--check 1`, so maybe you meant removing that and just having them under `tests/basic` or something like that?

**Comment by DanGrayson** [comment]
Yes, maybe just stop incorporating those tests in the binary, and remove the option.  If I don't remember using it ever, how useful can it have been?

And the end-user has gazillions of tests to run: each package can be checked.

**Comment by mahrud** [comment]
> By the way, I had to revert one of your commits in commit [af86e51](https://github.com/Macaulay2/M2/commit/af86e51d499aed8a82966fd9b617419fa201dc2a)

Wish you had made an issue of this first, to give me a chance to fix it before you revert it.

I can't reproduce the problem you mention in the commit message. Can you explain it?

**Comment by DanGrayson** [comment]
> > By the way, I had to revert one of your commits in commit [af86e51](https://github.com/Macaulay2/M2/commit/af86e51d499aed8a82966fd9b617419fa201dc2a)
> 
> Wish you had made an issue of this first, to give me a chance to fix it before you revert it.

Couldn't do that, as it was blocking development.

> 
> I can't reproduce the problem you mention in the commit message. Can you explain it?

The problem is that `Macaulay2/Core/nets.m2` and `Macaulay2/m2/nets.m2` are two names for the same file (because "Core" is a symbolic link to "m2"), and only one is recognized as being under source code control.  It's confusing to humans, too, so a solution without a symbolic link is necessary, even if emacs can be taught to handle it.

Why not rename the directory called "Core" in the distribution being assembled under `usr-dist` to "m2"?  Would that help?

**Comment by mahrud** [comment]
> The problem is that Macaulay2/Core/nets.m2 and Macaulay2/m2/nets.m2 are two names for the same file (because "Core" is a symbolic link to "m2"), and only one is recognized as being under source code control.

What did you do to get that error? My emacs opens files in Core without an issue. I guess my question is, are you sure this wasn't an issue with your setup?

> Why not rename the directory called "Core" in the distribution being assembled under usr-dist to "m2"? Would that help?

Since the package name is "Core", it would make more sense to rename "m2" in github to "Core".

**Comment by DanGrayson** [comment]
I'm against that idea, because it screws up the git history to rename files or directories.  And yes, there is a package
called Core, but it's not like the other packages, because the function newPackage is defined within it.  So one doesn't have to take that literally.

**Comment by mahrud** [comment]
> I'm against that idea, because it screws up the git history to rename files or directories.

What does it exactly break? `git blame` works without out of the box and `git log` works with the `--follow` workaround. I just tried both by looking at the history of `Macaulay2/tests/normal`, which used to be under `Macaulay2Doc`.

**Comment by mahrud** [comment]
Here's blame of a random test, showing commits from 14 years ago: https://github.com/Macaulay2/M2/blame/master/M2/Macaulay2/tests/normal/4-res-tower-crash.m2

**Comment by mahrud** [comment]
Another option is to rename the directory in the "past". The idea is to go to a past commit, say, 10 years ago, rename the directory and add a symlink `m2 -> Core`, then rebase and force push to the branch. Then even `git log` would show 10 years worth of history without `--follow`. A similar trick could be used to get rid of the top level "M2" directory if we wanted.

If this sounds like too much headache, I'm happy to try to get the symlink option to work again. Can you explain how I can reproduce it?

**Comment by DanGrayson** [comment]
No, let's not edit the history.  Just find a solution without a symbolic link.
